### PR TITLE
replicate dml statements in binlog to oracle

### PR DESCRIPTION
**Pull request by cartersz**
### What problem does this PR solve? 
replicate dml statements in binlog to oracle database

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: pkg/loader/model.go [hunk 1]
```diff
@@ -250,6 +300,62 @@ func (dml *DML) deleteSQL() (sql string, args []interface{}) {
...
+			colValues = colValues[:0]
+		}
+	}
+	// Fallback to use all columns
+	if len(colNames) == 0 {
+		for _, col := range dml.columnNames() {
+			colNames = append(colNames, col)
+			colValues = append(colValues, valueMap[col])
+		}
+	}
```
Comment (lichunzhu) [inline pkg/loader/model.go:342]: Why not use `whereSlice` here? This for loop will return **ALL** unique index columns.

File: pkg/loader/model_test.go [hunk 2]
```diff
@@ -262,3 +265,195 @@ func (s *SQLSuite) TestUpdateMarkSQL(c *check.C) {
...
 }
+
+func (s *SQLSuite) TestOracleUpdateSQL(c *check.C) {
+	dml := DML{
+		Tp:       UpdateDMLType,
+		Database: "db",
+		Table:    "tbl",
+		Values: map[string]interface{}{
+			"id":123,
+			"name": "pc",
```
Comment (lichunzhu) [inline pkg/loader/model_test.go:276]: Should use uppercase names.

File: pkg/loader/executor.go [hunk 3]
```diff
@@ -17,6 +17,7 @@ import (
 	"context"
 	gosql "database/sql"
 	"fmt"
+	"github.com/godror/godror"
```
Comment (lichunzhu) [inline pkg/loader/executor.go:20]: We should add some unit tests for oracle part in this file. Especially for the safe mode part.

File: pkg/loader/util.go [hunk 4]
```diff
@@ -171,6 +200,37 @@ func CreateDB(user string, password string, host string, port int, tls *tls.Conf
...
+	loc, err := time.LoadLocation("Local")
+	if err != nil {
+		return nil, err
+	}
+	pConnectString := ""
+	if connectString == "" {
+		if serviceName == "" {
+			return nil, errors.New("service-name should not be empty")
+		}
+		pConnectString = fmt.Sprintf("%s:%d/%s?connect_timeout=2", host, port, serviceName)
```
Comment (lichunzhu) [inline pkg/loader/util.go:214]: Why `connect_timeout=2`?

File: pkg/loader/util.go [hunk 5]
```diff
@@ -259,6 +319,36 @@ func getColsOfTbl(db *gosql.DB, schema, table string) ([]string, error) {
...
 
+func getOracleColsOfTbl(db *gosql.DB, schema, table string) ([]string, error) {
+	rows, err := db.Query(colsOracleSQL, schema, table)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer rows.Close()
+	cols := make([]string, 0, 1)
+	for rows.Next() {
+		var name, virtualColumn string
```
Comment (lichunzhu) [inline pkg/loader/util.go:330]: Why do we need to select `virtualColumn`? It's never used.

File: pkg/loader/util.go (hunk 4 above)
Comment (cartersz) [reply pkg/loader/util.go:214]: timeout time of connecting to db suggested by dba

File: pkg/loader/util.go (hunk 5 above)
Comment (cartersz) [reply pkg/loader/util.go:330]: yes, it can be deleted.

File: drainer/translator/oracle.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+)
+
+// TiBinlogToOracleTxn translate the format to loader.Txn
+func TiBinlogToOracleTxn(infoGetter TableInfoGetter, schema string, table string, tiBinlog *tipb.Binlog, pv *tipb.PrewriteValue, shouldSkip bool, tableRouter *router.Table) (txn *loader.Txn, err error) {
+	txn = new(loader.Txn)
+
+	if tiBinlog.DdlJobId > 0 {
+		downStreamSchema, downStreamTable, routeErr := tableRouter.Route(schema, table)
+		if routeErr != nil {
+			return nil, errors.Annotatef(routeErr, "when binlog to oracle ddl txn,route schema and table failed. schema=%s, table=%s", schema, table)
```
Comment (lichunzhu) [inline drainer/translator/oracle.go:21]:
```suggestion
			return nil, errors.Annotatef(routeErr, "when binlog to oracle ddl txn, route schema and table failed. schema=%s, table=%s", schema, table)
```

File: drainer/translator/oracle.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+			canAppendDefaultValue := infoGetter.CanAppendDefaultValue(mut.GetTableId(), pv.SchemaVersion)
+
+			schema, table, ok = infoGetter.SchemaAndTableName(mut.GetTableId())
+			if !ok {
+				return nil, errors.Errorf("SchemaAndTableName empty table id: %d", mut.GetTableId())
+			}
+			downStreamSchema, downStreamTable, routeErr := tableRouter.Route(schema, table)
+			if routeErr != nil {
+				return nil, errors.Annotate(routeErr, fmt.Sprintf("when binlog to oracle dml txn,route schema and table failed. schema=%s, tabel=%s", schema, table))
```
Comment (lichunzhu) [inline drainer/translator/oracle.go:54]: ditto

File: pkg/loader/executor.go [hunk 8]
```diff
@@ -302,15 +358,28 @@ func tryRefreshTableErr(err error) bool {
 	return false
 }
 
+func tryRefreshTableOracleErr(err error) bool {
+	oraErr, ok := godror.AsOraErr(err)
+	if !ok {
+		return false
+	}
+	//Invalid identifier for oracle db error
```
Comment (lichunzhu) [inline pkg/loader/executor.go:366]: Is there any link for this code?
```suggestion
	// Invalid identifier for oracle db error
```

File: pkg/loader/executor.go [hunk 9]
```diff
@@ -250,19 +294,31 @@ func (e *executor) execTableBatch(ctx context.Context, dmls []*DML) error {
...
+		if err := e.splitExecDML(ctx, allInserts, bulkInsert); err != nil {
 			return errors.Trace(err)
 		}
 	}
 
 	if allUpdates, ok := types[UpdateDMLType]; ok {
-		if err := e.splitExecDML(ctx, allUpdates, e.bulkReplace); err != nil {
+		bulkUpdate := e.bulkReplace
+		if e.destDBType == "oracle" {
+			bulkUpdate = e.oracleBulkOperation
```
Comment (lichunzhu) [inline pkg/loader/executor.go:319]: Does oracle have bulk operation like MySQL? `e.oracleBulkOperation` is actually not a bulk execution.
For example:
```sql
insert into xx (aa, bb) values
(1,1),
(2,2),
(3,3);
```

File: pkg/loader/load.go [hunk 10]
```diff
@@ -724,6 +743,7 @@ type batchManager struct {
 	fDMLsSuccessCallback func(...*Txn)
 	fExecDDL             func(*DDL) error
 	fDDLSuccessCallback  func(*Txn)
+	destDBType           string
```
Comment (lichunzhu) [inline pkg/loader/load.go:746]: This variable is never used.
```suggestion
```

File: pkg/loader/load_test.go [hunk 11]
```diff
@@ -384,13 +384,43 @@ func (s *execDDLSuite) TestShouldUseDatabase(c *check.C) {
...
+	mock.ExpectBegin()
+	mock.ExpectExec("BEGIN test.do_truncate").WillReturnResult(sqlmock.NewResult(0, 0))
+	mock.ExpectCommit()
+
+	loader := &loaderImpl{db: db, ctx: context.Background(), destDBType: "oracle"}
+
+	ddl := DDL{SQL: "alter table t1 truncate partition p1", Database: "test", Table: "t1"}
+	err = loader.execDDL(&ddl)
+	c.Assert(err, check.IsNil)
+}
```
Comment (lichunzhu) [inline pkg/loader/load_test.go:422]: What's these two tests 🤔

File: pkg/loader/model_test.go [hunk 12]
```diff
@@ -262,3 +265,300 @@ func (s *SQLSuite) TestUpdateMarkSQL(c *check.C) {
...
+			columns: []string{"ID", "ID2", "NAME", "C2"},
+			uniqueKeys: []indexInfo{
+				{
+					name:    "uniq_index_name_id2",
+					columns: []string{"ID2"},
+				},
+				{
+					name:    "uniq_index_name_id",
+					columns: []string{"ID"},
+				},
```
Comment (lichunzhu) [inline pkg/loader/model_test.go:513]: make sure not picking a unique index with nil value
```suggestion
				{
					name:    "uniq_index_name_id",
					columns: []string{"ID", "C2"},
				},
				{
					name:    "uniq_index_name_id2",
					columns: []string{"ID2"},
				},
```

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

**Comment by lichunzhu** [comment]
@lance6716 @glorv PTAL

File: drainer/util.go [hunk 13]
```diff
@@ -262,20 +262,35 @@ func genRouterAndBinlogEvent(cfg *SyncerConfig) (*router.Table, *bf.BinlogEvent,
...
+	// only support two type ddl[truncate table xxx, and alter table xx truncate partition xx] for oracle db
+	if cfg.DestDBType == "oracle" {
+		filterRules = append([]*bf.BinlogEventRule{{
+			Action:        bf.Do,
+			SchemaPattern: "*",
+			TablePattern:  "*",
+			Events:        []bf.EventType{bf.TruncateTable, bf.AlertTable, bf.AllDML},
+			SQLPattern:    []string{".*truncate table.*", ".*alter table.*truncate partition.*"},
+		}}, filterRules...)
+	}
```
[the remark this answers is not in the document]
Comment (cartersz) [reply drainer/util.go]: this change will make a bug. I have test it. So this suggestion can not apply.

File: pkg/loader/load_test.go (hunk 11 above)
Comment (cartersz) [reply pkg/loader/load_test.go:422]: one for "truncate table xx" statement, another is for "alter table xx truncate partition xx"

File: pkg/loader/executor.go (hunk 9 above)
Comment (cartersz) [reply pkg/loader/executor.go:319]: DBA will give me some statements like mysql on next version.

File: pkg/loader/load.go (hunk 10 above)
Comment (cartersz) [reply pkg/loader/load.go:746]: yes

Comment (cartersz) [reply pkg/loader/load.go:746]: yes, it can be deleted.

File: pkg/loader/load_test.go (hunk 11 above)
Comment (lichunzhu) [reply pkg/loader/load_test.go:422]: We don't have DDL part now and these two ddls look different. I think we can add these tests later.

File: pkg/loader/model.go (hunk 1 above)
Comment (cartersz) [reply pkg/loader/model.go]: it is not suitable for my case

File: pkg/loader/executor.go (hunk 9 above)
Comment (lichunzhu) [reply pkg/loader/executor.go:319]: ok, can we left a TODO here?
```
			// TODO: add bulkOperation for oracle
			bulkUpdate = e.oracleBulkOperation
```

File: pkg/loader/load_test.go (hunk 11 above)
Comment (cartersz) [reply pkg/loader/load_test.go:422]: OK

File: pkg/loader/executor.go (hunk 9 above)
Comment (cartersz) [reply pkg/loader/executor.go:319]: yes

**Review by lichunzhu — COMMENTED** [review]
LGTM, but please fix the unit test and integration tests

**Review by lichunzhu — APPROVED** [review]
LGTM

**Review by glorv — COMMENTED** [review]
rest LGTM

File: drainer/sync/oracle.go [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+// Close implements Syncer interface
+func (m *OracleSyncer) Close() error {
+	m.loader.Close()
+
+	err := <-m.Error()
+
+	if m.relayer != nil {
+		closeRelayerErr := m.relayer.Close()
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (cartersz) [reply drainer/sync/oracle.go:105]: done

**Review by glorv — APPROVED** [review]

**Pull request merged**